General support for auxiliary resources

joachimvh opened the pull request:
This is a lot of text (after re-reading this it got a bit longer than expected 😅), but several things in this PR are not set in stone yet so the discussion is the core part. This PR is partially just to see how it could work.

Suggestion on how to handle auxiliary files in a more general way. Currently we only have acl files, but the option exists for more and at some point we at least want to add shape files. .meta files are currently handled differently but this could also change.

There are several design decisions here where there were multiple options which I'll go over below as we might want to change some. The main functions are mostly the same as those that were in AclManager. New ones are:
 * getRequiredPermissions -> Currently used because acl files require different permissions. Could also be removed by editing the authorizer (more on this below)
 * getType -> Returns the type of this auxiliary resource. Used for adding metadata needed for the link headers and could also be used for identifying what type of auxiliary resource we have if needed
 * validate -> used to check if an auxiliary resource is valid. Currently this just checks if the data can be converted (so if it's valid RDF), but this could be extended in the future with shape validation for example.

Some more notes about the interface:
 * All functions take a ResourceIdentifier as input. Reason for this is because otherwise it would be impossible to make the `CompositeAuxiliaryManager` which I think could be useful for when we have multiple types. Although this could be thinking too far ahead.
 * Wasn't really sure where to put everything, it's currently in ldp/auxiliary.
 * Removed some promises from the original AclManager functions as chances are low those would ever be relevant.

Besides that interface the acl authorizer has been updated. Before, the authorizer had a single if-statement in the beginning to check if the input was an acl file and to take the correct permissions on that. This has now been removed to make the authorizer more consistent. Instead, a new Authorizer has been added for auxiliary resources (and added in a waterfall handler). In case the input is an auxiliary resource the permissions will be changed (as mentioned above), and the request will be passed to the default authorizer with the URI of the associated resource instead. Instead of having `getRequiredPermissions` in the auxiliary manager interface, it would also be possible to instead have an exception for acl files in this class (either by having 2 manager, or using `getType` to check what kind of auxiliary resource it is). Later edit: probably should make a specific permissions extractor for acl files, would make more sense than all the other suggestions.

There are still some more things to be added but I already wanted to create this PR for discussion and in case there are problems.

To be added:
 * A new LockingResourceStore. The new one would extend the old one, but edit the calls so the identifier of the associated resource is used for locking when performing operations on auxiliary resources.
 * A new ResourceStore to handle auxiliary behaviour, including:
   * Validation checks on all incoming auxiliary data.
   * Deleting auxiliary resources when the main resource gets deleted. Although other solutions were also suggested in #53, I would like to keep everything auxiliary related out of the `DataAccessorBasedStore` if possible.
   * Add necessary metadata when getting the data (to create the link headers). Writing this just made me realize that the CompositeAuxiliaryHandler is not going to work as I wanted it to (since you need types for all auxiliary resources), so should probably scrap that one again.
   * Specifically for acl files we want to make sure they can't be deleted if they're in a root storage container (#381). Not sure how to best do this one, either as a specific check in this store or a separate store just for this. Or perhaps we event want a separate store for every auxiliary type instead of having 1 store that handles all types. We probably never are going to have so many auxiliary types that this would become problematic so I'm sort of a fan of this idea.
   * Potentially filter out the references to auxiliary resources when doing a GET request on containers. Currently this would still be a bit annoying though, depending on #384.

After last changes, now closes #53, closes #381, closes #385 and closes #573.

joshdcollins commented:
I've been following CSS' development, with a specific interest in how auxiliary resources would be supported.

Apologies in advance of any of these questions / thoughts are premature given the state of the PR...

1. It looks like if I wanted to add support for a new aux resource the following changes would be required:
   - Create an implementation of `SuffixAuxiliaryManager` to determine the suffix to use
   - Create an implementation of `MetadataWriter` to express the link header
   - Wire the above into the config JSON

Does that cover it or would other steps be missing?

2. It would be a welcome developer / configurer experience to be able to define "generic" auxiliary resources with only changes in the config JSON by specifying suffix and required permissions without the need to create new classes (unless extending functionality was required).  It appears that the `SuffixAuxiliaryManager` is very close to that (or perhaps accomplishes it now and I don't fully grasp the DI of component.js).

Thanks!
jc

joachimvh commented:
> Apologies in advance of any of these questions / thoughts are premature given the state of the PR...

Like you see, this PR is not a finished product yet and more of a look into how it could be done. After pushing this I already had some different ideas about how to tackle some things here. So my answer here is going to be more of a thought experiment on "what could be" once this PR would be finished.

>  1. It looks like if I wanted to add support for a new aux resource the following changes would be required:
>        
>        * Create an implementation of `SuffixAuxiliaryManager` to determine the suffix to use
>        * Create an implementation of `MetadataWriter` to express the link header
>        * Wire the above into the config JSON
> 
>  Does that cover it or would other steps be missing?

It depends on how similar the wanted behaviour is for your new type of aux resource compared to those that already exist. The main idea behind this PR is that the few aux resources that are defined in the spec (acl, shape, potentially meta) share several behaviours:
 * They have to be deleted when the main resource is deleted.
 * Permissions have to be checked on the main resource when accessing them.
 * They have to be exposed through a link header.
 * We want to not include them when showing the contents of a container.
 * Internally we want our locking system to lock the main resource when accessing them.
 * Their data has to be validated before being stored.
 * Potentially something I'm forgetting.
In case your aux resource wants different behaviour it would need a custom implementation, so let's assume that it does want this behaviour.

An implementation of `AuxiliaryManager` would be needed then. `SuffixAuxiliaryManager` is a helper class to cover aux files with even more similar behaviour: their URIs are made by adding a suffix (e.g., `.acl`) to the original URL, and they have to at least be valid RDF.

Writing a `MetadataWriter` would actually not be needed. Currently there's one in place for acl, but the idea would be to remove that one once this PR is finished. As mentioned in the "to be added" section above, the idea would then be to have a class that, for all aux types, handles adding the correct metadata. We already have a generic metadata writer in place that can convert this kind of metadata to a correct header. So the only thing that would be required there would be some extra configuration in the JSON.

>   2. It would be a welcome developer / configurer experience to be able to define "generic" auxiliary resources with only changes in the config JSON by specifying suffix and required permissions without the need to create new classes (unless extending functionality was required).  It appears that the `SuffixAuxiliaryManager` is very close to that (or perhaps accomplishes it now and I don't fully grasp the DI of component.js).

That is true actually. Currently it would not be possible due to being an abstract class, but it would not require that much change to be an actual class that can completely be configured. The main reason it's currently not like this is because the 2 main cases I considered are those from the spec (acl and shapes), and both of these would require custom validation anyway besides just the RDF check. But if there is a use case for supporting new aux types that only require RDF validation that is certainly something we could consider.

joshdcollins commented:
Thanks for the response -- I appreciate the detail.

In thinking more about it, I was likely over-simplifying the lifecycle of aux resources and agree that the case for purely configurable ones is probably more narrow than I had thought.

I could imagine a use case for purely "client-managed" aux resources that have the behavior you bulleted above.  For instance, in advance of server-side support for [shape trees](https://shapetrees.github.io/specification/spec#discover) we are performing client-side validation and just requiring that a server be able to expose a given link header and store content associated with it.  That said, implementing this would be trivial given your proposal above.

Thanks again!

RubenVerborgh commented on config/presets/acl.json:
```diff
@@ -3,10 +3,33 @@
     },
+
     {
       "@id": "urn:solid-server:default:AclAuthorizer",
```
Perhaps for clarity now
```suggestion
      "@id": "urn:solid-server:default:AclBasedAuthorizer",
```

RubenVerborgh commented on config/presets/acl.json:
```diff
@@ -3,10 +3,33 @@
+      "@type": "WaterfallHandler",
+      "WaterfallHandler:_handlers": [
+        {
+          "@type": "AuxiliaryAuthorizer",
```
Perhaps insert a comment

RubenVerborgh commented on src/authorization/AuxiliaryAuthorizer.ts:
```diff
@@ -0,0 +1,55 @@
+import { Authorizer } from './Authorizer';
+
+/**
+ * An authorizer for auxiliary resources.
```
```suggestion
 * An authorizer for auxiliary resources such as …
```

RubenVerborgh commented on src/authorization/AuxiliaryAuthorizer.ts:
```diff
@@ -0,0 +1,55 @@
+
+/**
+ * An authorizer for auxiliary resources.
+ * Calls the source authorizer with potentially updated permissions
```
Perhaps add some explanation of how permissions for auxiliary resources actually depend on those on other resource, otherwise this is hard to understand.

RubenVerborgh commented on src/authorization/AuxiliaryAuthorizer.ts:
```diff
@@ -0,0 +1,55 @@
+  private readonly authorizer: Authorizer;
+  private readonly manager: AuxiliaryManager;
+
+  public constructor(authorizer: Authorizer, manager: AuxiliaryManager) {
```
If you think it helps clarity (also for local var).

```suggestion
  public constructor(resourceAuthorizer: Authorizer, manager: AuxiliaryManager) {
```

RubenVerborgh commented on src/authorization/AuxiliaryAuthorizer.ts:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  private getSafeAssociatedArgs(input: AuthorizerArgs): AuthorizerArgs {
+    if (!this.manager.isAuxiliaryIdentifier(input.identifier)) {
+      throw new NotImplementedHttpError('AuxiliaryAuthorizer only supports auxiliary resources.');
+    }
```
Let's just merge this into the function below; splitting them is premature optimization.

RubenVerborgh commented on src/authorization/AuxiliaryAuthorizer.ts:
```diff
@@ -0,0 +1,55 @@
+    return this.getAssociatedArgs(input);
+  }
+
+  private getAssociatedArgs(input: AuthorizerArgs): AuthorizerArgs {
```
```suggestion
  private getRequiredAuthorization(input: AuthorizerArgs): AuthorizerArgs {
```

RubenVerborgh commented on src/authorization/AuxiliaryAuthorizer.ts:
```diff
@@ -0,0 +1,55 @@
+    this.manager = manager;
+  }
+
+  public async canHandle(input: AuthorizerArgs): Promise<void> {
+    const newArgs = this.getSafeAssociatedArgs(input);
```
Just for clarity
```suggestion
  public async canHandle(auxiliaryAuthorization: AuthorizerArgs): Promise<void> {
    const resourceAuthorization = this.getRequiredAuthorization(auxiliaryAuthorization);
```

RubenVerborgh commented on src/authorization/AuxiliaryAuthorizer.ts:
```diff
@@ -0,0 +1,55 @@
+
+  public async handle(input: AuthorizerArgs): Promise<void> {
+    const newArgs = this.getAssociatedArgs(input);
+    this.logger.debug(`Redirecting auth request for ${input.identifier.path} to ${newArgs.identifier.path}`);
```
To avoid HTTP redirection confusion:
```suggestion
    this.logger.debug(`Checking authorization for ${input.identifier.path} on ${newArgs.identifier.path}`);
```

RubenVerborgh commented on src/authorization/WebAclAuthorizer.ts:
```diff
@@ -24,17 +25,24 @@ import { Authorizer } from './Authorizer';
   protected readonly logger = getLoggerFor(this);
 
-  private readonly aclManager: AclManager;
+  private readonly aclManager: AuxiliaryManager;
```
Rename var too?

joachimvh replied:
I actually kept this name on purpose to indicate that we specifically have an auxiliary manager for acl files here.

RubenVerborgh commented on src/authorization/WebAclAuthorizer.ts:
```diff
@@ -24,17 +25,24 @@ import { Authorizer } from './Authorizer';
 
+  public async canHandle({ identifier }: AuthorizerArgs): Promise<void> {
+    if (this.aclManager.isAuxiliaryIdentifier(identifier)) {
+      throw new NotImplementedHttpError('WebAclAuthorizer does not support permissions on acl files.');
```
```suggestion
      throw new NotImplementedHttpError('WebAclAuthorizer does not support permissions on auxiliary resources.');
```

joachimvh replied:
The auxiliary manager it takes as input is one specifically for acl files (which it needs to find the correct acl resource for a permission), so the check here will only check specifically for acl resources and not all auxiliary resources in general. "files" still needs to be changed to "resources" though.

RubenVerborgh commented on src/authorization/WebAclAuthorizer.ts:
```diff
@@ -132,7 +138,7 @@ export class WebAclAuthorizer extends Authorizer {
    * Returns the acl triples that are relevant for the given identifier.
    * These can either be from a corresponding acl file or an acl file higher up with defaults.
-   * Rethrows any non-NotFoundHttpErrors thrown by the AclManager or ResourceStore.
+   * Rethrows any non-NotFoundHttpErrors thrown by the aclManager or ResourceStore.
```
?

joachimvh replied:
I probably did a replace-all for something and this was caught in the crossfire 😅

joachimvh replied:
Ah no, I just remembered. I actually did this on purpose since there no longer is an AclManager class but only an aclManager variable.

RubenVerborgh commented on src/ldp/auxiliary/UrlBasedAclManager.ts:
```diff
@@ -0,0 +1,35 @@
+ */
+export class UrlBasedAclManager extends SuffixAuxiliaryManager {
+  public constructor(converter: RepresentationConverter) {
+    super('.acl', converter);
```
Configurable?

RubenVerborgh commented on src/ldp/http/metadata/AclLinkMetadataWriter.ts:
```diff
@@ -9,17 +9,17 @@ import { MetadataWriter } from './MetadataWriter';
  */
 export class AclLinkMetadataWriter extends MetadataWriter {
-  private readonly aclManager: AclManager;
+  private readonly aclManager: AuxiliaryManager;
```
name?

joachimvh replied:
Same reason as above.

RubenVerborgh commented on src/ldp/permissions/MethodPermissionsExtractor.ts:
```diff
@@ -21,6 +21,7 @@ export class MethodPermissionsExtractor extends PermissionsExtractor {
     const read = READ_METHODS.has(method);
     const write = WRITE_METHODS.has(method);
-    return { read, write, append: write };
+    const control = false;
+    return { read, write, append: write, control };
```
```suggestion
    return { read, write, append: write, control: false };
```
for consistency (or the other way count, make `append` a `const`)

RubenVerborgh reviewed (approved):
Agree with conceptual part and implementation. Well done 👍👍👍

joachimvh commented:
This PR has all the planned features now and closes 4 issues (added them in main message for auto-closure). Since it's quite big I'm going to link changes done to commits.

Changes since last version:
 * All resolved comments above
 * `getType` was replaced with `addMetadata`
 * SuffixAuxiliaryManager takes an optional `link` parameter to be used for adding metadata
 * `getPermissions` was removed, there now is an acl permissions extractor instead
 * There is a new function in the interface: `canDeleteRoot` (more on this one in "Things I don't like")
 * UrlBasedAclManager is now almost empty but we will need this in the future anyway for acl validation
 * CompositeAuxiliaryManager now has a function `getAuxiliaryIdentifiers` (also more on this in "Things I don't like")
 * Updated some RepresentationMetadata helper functions (needed for commit after that)
 * DataAccessorBasedStore now does the spec required auxiliary actions
 * AclLinkMetadataWriter was removed since it can now be simply configured

### Changes to DataAccessorBasedStore
This was the main goal of this PR, as this causes us to support most of the spec requirements regarding auxiliary resources.
The following features were added to the store:
 * GET: add auxiliary-related metadata
 * GET: remove ldp:contains triples of auxiliary resources
 * POST: reject slugs that would create an auxiliary resource
 * PUT: validate the incoming data of auxiliary resources
 * PATCH: no change here of course, but because of the other changes PATCH requests will also get validated (#571)
 * DELETE: prevent deletion of acl files in root storage containers
 * DELETE: allow deletion of containers with only auxiliary resources
 * DELETE: also delete corresponding auxiliary resources when deleting a normal resource

I initially tried creating a separate store with this functionality, but supporting all of this was just too much of a hassle. Especially the DELETE additions would require so many more calls to the backend store to be supported correctly.

This does tie this logic into the DataAccessorBasedStore, but on the other hand, the DataAccessorBasedStore already was our central point of supporting Solid logic on resources: not allowing deletion of root storage containers, preventings POSTs to non-containers, not allowing identifiers with and without trailing slash, and many many more. In this way the advantage is that there still is only one place to look for most of the spec required behaviour regarding resources.

The main issue with this is if in the future someone wants to create a new store (not a new DataAccessor). That store would then have to replicate all of this behaviour. But this would already have been a problem anyway, this PR does not increase the complexity of that problem that much. Potentially this could be easily fixed by moving more parts of the behaviour to separate protected functions in the DataAccessorBasedStore so they can be reused. Another solution would be to completely rethink the architecture of this class so functionality can be abstracted to separate classes that can be plugged in? I would say to cross that bridge when we get there, although if we do want to take this situation into account perhaps we can make an issue around this and try to make a new backend ResourceStore and see what is missing to make that easy.

### Things I don't like
As mentioned above, there are 2 things in this PR I'm not really happy with and am open to suggestions on how to change those parts.

One is the fact that that acl files in root storage containers can't be deleted. For this the AuxiliaryManager interface now has the function `canDeleteRoot` which returns a boolean. But the only point of this function is for this specific case so it feels a bit unfortunate to add it just for that. On the other hand I have no idea how to improve upon this. Other solutions could be to add a separate store that supports just this behaviour, or have the DataAccessorBasedStore take 2 separate auxiliary managers as input, but both of those feel way worse

The other one is the `getAuxiliaryIdentifiers` function in the composite auxiliary manager. This is needed so we can know all auxiliary resources related to the main resource, so we can delete them together with the main resource. But this now causes the composite manager to expose a function not in the interface, so the DataAccessorBasedStore is required to take that specific type of manager as input parameter instead of a general auxiliary manager. Some other solutions (for each of which a case can be made imo):
 * Delete the CompositeAuxiliaryManager and have the store take an array of auxiliary managers instead. The disadvantage of this is that we would then need helper functions for all the things the composite manager does. There are more cases where we need to handle all auxiliary types separately so would need several new methods to iterate over them (either in the store or a separate AuxiliaryUtil file)
 * Add the `getAuxiliaryIdentifiers` function to the AuxiliaryManager. This might be the cleanest solution but on the other hand for single auxiliary managers it would be silly to have this function.

In comparison to the rest of the PR these are 2 minor things but they're the main ugly things still in there so I would be happy to have a cleaner solution there.

joshdcollins commented:
@joachimvh As a point of validation for the changes in this PR -- I was able to fairly trivially implement support for exposing and maintaining an additional aux. resource -- https://github.com/janeirodigital/community-server/commit/961f99f6fc43a598307f07741d18deb9ccd05290

RubenVerborgh commented on src/ldp/auxiliary/AuxiliaryManager.ts:
```diff
@@ -0,0 +1,70 @@
+  isAuxiliaryIdentifier: (identifier: ResourceIdentifier) => boolean;
+
+  /**
+   * Returns the identifier of the resource which this auxiliary resource is referencing to.
```
```suggestion
   * Returns the identifier of the resource which this auxiliary resource is referring to.
```

RubenVerborgh reviewed (changes requested):
Thanks for the hard work and the detailed explanation, @joachimvh.
After reviewing the code, I think this is the way to go, and I have some comments that will hopefully lead to further simplifications.

It is important to note that the architecture upon which the server is based, was created before the need for multiple auxiliary resources was apparent. So the friction we feel is because the architecture indeed did not take this into account. However, by extending the `ResourceStore` contract to also cover auxiliary resources (as we do here), I think we're there.

---

- [ ] Let's change the `ResourceStore` interface TSdoc such that the contract becomes that it also manages auxiliary resources.

---

> * Add the `getAuxiliaryIdentifiers` function to the AuxiliaryManager.

- [ ] This is the way to go indeed. (However, comment below).

---

I notice that currently, aux managers are doing three things:
1. identifier management
2. adding metadata
3. validation

And `canDeleteRoot` too. But let's tackle that one later.

However, those 3 functions are independent—yet they are currently tied into a manager.

Concretely: ACL resources have `.acl` and need RDF validation.
But even if ACL resources would follow an `?acl` pattern (as ESS currently does), they would still need validation.
And the same metadata.

And non-ACL auxiliaries might also need the RDF validation, even if they are not suffix-based!

As per your explanation above, I note that the existence of a nearly empty `UrlBasedAclManager` is because you are leaving an opening for validation. And similarly, @joshdcollins has `ShapeTreeAuxiliaryManager`. But I don't think you have to create subclasses at all—this can all done via configuration.

- [ ] Step 1 is to split validation out of the aux managers. They should be independent validators.

- [ ] Then we don't need to create `UrlBasedAclManager` and `ShapeTreeAuxiliaryManager`; they both are simply a (Components.js) configuration of `AuxiliaryManager`, where you pass the required validator (possibly composite: RDF syntax + further checks) as a constructor argument, as well as `".acl"` for the extension and `false` for `canDeleteRoot`.

- [ ] But I think we can even go further: rename `AuxiliaryManager` into `AuxiliaryStrategy` (it's not managing anything itself) and make it a composite component that is nothing more than an aggregation of three objects:
- `identifiers: IdentifierStrategy` (or `IdentifierManager`)
  - `isAuxiliaryIdentifier`
  - `getAuxiliaryIdentifiers`
  - `getAssociatedIdentifier`
- `metadataGenerator: AsyncHandler<{ identifier: ResourceIdentifier, metadata: RepresentationMetadata }, void>`
  - `handle` (implements `addMetadata`)
- `validator: : AsyncHandler<{ identifier: ResourceIdentifier, metadata: RepresentationMetadata }, void>`
  - `handle` (implements `validate`)

- [ ] In the last two cases, I wonder if we even need to pass `identifier`, or whether `metadata` covers that.

- [ ] Some components, such as `AuxiliaryLockingStore`, can only take the `IdentifierStrategy` bit; they don't need the whole object.

- [ ] We _could_ also make the `canDeleteRoot` method part of this. However, I propose to simply drop it. I conjecture it will always be `false`. If we need it at any point, we introduce it and set its value through dependency injection.

joachimvh commented:
(Replying to the main comment before going through the individual comments)

It's the main problem of how much modularization we want vs how complicated it would become to configure everything. Will look into doing it like this since I don't see any major problems with this at first.

> However, those 3 functions are independent—yet they are currently tied into a manager.

Don't completely agree since the identifier management is probably always going to be necessary for adding the metadata. They can still be separated though but the metadata would need to be injected with the identifier management. Or the IdentifierStrategy could be a required parameter for the handle method of the metadata generator. Potentially also for the validator once we do more than simple RDF checking there.

> In the last two cases, I wonder if we even need to pass identifier, or whether metadata covers that.

Yea I also doubted there. I kept identifier to be in line with the other functions but they could probably be removed.

> We could also make the canDeleteRoot method part of this. However, I propose to simply drop it. I conjecture it will always be false.

There are currently 2 examples of auxiliary resources in the spec: acl files and description resources. We currently obfuscate the second one by internally storing it as metadata, but that is an example of an auxiliary resouce needing this value to be `true`.

joachimvh commented on src/ldp/auxiliary/CompositeAuxiliaryManager.ts:
```diff
@@ -0,0 +1,70 @@
+ * An {@link AuxiliaryManager} that combines multiple AuxiliaryManagers into one.
+ * Uses `isAuxiliaryIdentifier` to know which manager to call for which call.
+ */
+export class CompositeAuxiliaryManager implements AuxiliaryManager {
```
There is something to be said for generalizing composite classes, but I'm going to skip that for this PR.

joachimvh commented on src/ldp/auxiliary/SuffixAuxiliaryManager.ts:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  public async validate(identifier: ResourceIdentifier, representation: Representation): Promise<void> {
+    // Read data in memory first so it does not get lost
+    const data = await arrayifyStream(representation.data);
+    const preferences = { type: { [INTERNAL_QUADS]: 1 }};
+
+    // Creating new representation since converter might edit metadata
+    const dummyRepresentation = new BasicRepresentation(data, identifier, representation.metadata.contentType);
```
Yes, the only thing this function does is check if this is valid RDF, which it does by trying to convert the entire data stream to quads. We don't want to change anything to the original data stream though, which is why everything is buffered beforehand.

joachimvh commented on src/ldp/auxiliary/SuffixAuxiliaryManager.ts:
```diff
@@ -0,0 +1,83 @@
+    // Stream has been drained so need to create new stream
+    representation.data = guardedStreamFrom(data);
+  }
+}
```
> Does this assume that all aux files are RDF?
When I originally wrote the idea for this class I saw all examples of auxiliary files were RDF so I made a helper function for that, but now I see that the spec actually has been updated to say that all auxiliary files have to be RDF ( "Auxiliary resources are represented as RDF documents" ), so now we have to assume that.

joachimvh commented on src/storage/AuxiliaryLockingStore.ts:
```diff
@@ -0,0 +1,57 @@
+ * when the input is an auxiliary resource.
+ * The request itself remains unchanged and is still executed on the auxiliary resource.
+ */
+export class AuxiliaryLockingStore extends LockingResourceStore {
```
Either way works for me so I'm ok with changing this. I wrote this before I started on the DataAccessorBasedStore so then I still tried to keep auxiliary as separate as possible. Will change it to one class for consistency.

joachimvh commented on src/ldp/auxiliary/SuffixAuxiliaryManager.ts:
```diff
@@ -0,0 +1,83 @@
+    // Creating new representation since converter might edit metadata
+    const dummyRepresentation = new BasicRepresentation(data, identifier, representation.metadata.contentType);
+    const result = await this.converter.handleSafe({ identifier, representation: dummyRepresentation, preferences });
+    // Drain stream to make sure data was parsed correctly
+    await new Promise((resolve, reject): void => {
+      result.data.on('data', (): void => {
+        // Drain the data
+      });
+      result.data.on('error', reject);
+      result.data.on('end', resolve);
+      result.data.on('close', resolve);
+    });
```
In hindsight I should have just used `arrayifyStream` which would do exactly the same.

joachimvh commented on src/ldp/permissions/AclPermissionsExtractor.ts:
```diff
@@ -0,0 +1,36 @@
+  }
+
+  public async handle(): Promise<PermissionSet> {
+    return {
+      read: false,
+      write: false,
+      append: false,
+      control: true,
+    };
```
Perhaps we want to keep it like this just in case some other class modifies this object? We can of course return `{ ...permissions }` but then we don't really win anything.

joachimvh commented:
Partially related: we already have something called an `IdentifierStrategy`, which mostly covers relative container behaviour: https://github.com/solid/community-server/blob/main/src/util/identifiers/IdentifierStrategy.ts

So this here we would need something like `AuxiliaryIdentifierStrategy`.

joachimvh commented:
Version 3!

Although there are some comments on the previous version that could change things, this should already mostly match the new suggestion.

Although a lot is re-written, functionality wise not much changed. The changes to the DataAccessorBasedStore are pretty much the same for example. The main changes are the auxiliary related interfaces which have completely changed, and the corresponding configs.

An overview (main points are interfaces, sub points are implementations):
  * `AuxiliaryIdentifierStrategy` has all the identifier related functions. Now with `getAuxiliaryIdentifiers` although it still sort of feels weird there. Also the name is very long.
    * `SuffixAuxiliaryIdentifierStrategy` generates identifiers based on a suffix
    * `RoutingAuxiliaryIdentifierStrategy` combines multiple strategies into one.
  * `MetadataGenerator` takes as input a RepresentationMetadata and adds to it somehow
    * `LinkMetadataGenerator` adds links to auxiliary resources based on a given link predicate
  * `Validator` takes a Representation as input and validates it somehow. There's something to be said for calling it a `RepresentationValidator`.
    * `RdfValidator` validates a representation by checking if the data stream contains valid RDF in any format supported by the given converter
  * `AuxiliaryStrategy` extends `AuxiliaryIdentifierStrategy` and adds `canDeleteRoot`, `addMetadata` and `validate` functions
    * `CompositeAuxiliaryStrategy` takes as input an `AuxiliaryIdentifierStrategy`, a `MetadataGenerator`, a `Validator` and a boolean (for canDeleteRoot)
    * `RoutingAuxiliaryStrategy` extends `RoutingAuxiliaryIdentifierStrategy` and adds support for the 3 other functions

An example of how the acl strategy is now configured:
```json
    {
      "@id": "urn:solid-server:default:AclIdentifierStrategy",
      "@type": "SuffixAuxiliaryIdentifierStrategy",
      "SuffixAuxiliaryIdentifierStrategy:_suffix": ".acl"
    },
    {
      "@id": "urn:solid-server:default:AclStrategy",
      "@type": "CompositeAuxiliaryStrategy",
      "CompositeAuxiliaryStrategy:_identifierStrategy": {
        "@id": "urn:solid-server:default:AclIdentifierStrategy"
      },
      "CompositeAuxiliaryStrategy:_metadataGenerator": {
        "@type": "LinkMetadataGenerator",
        "LinkMetadataGenerator:_link": "http://www.w3.org/ns/auth/acl#accessControl",
        "LinkMetadataGenerator:_identifierStrategy": {
          "@id": "urn:solid-server:default:AclIdentifierStrategy"
        }
      },
      "CompositeAuxiliaryStrategy:_validator": {
        "@id": "urn:solid-server:default:RdfValidator"
      },
      "CompositeAuxiliaryStrategy:_canDeleteRoot": false
    }
```
So some duplication on the `AclIdentifierStrategy` part. This entire block is what would be needed for a new auxiliary resource.

For all classes that need to support all auxiliary resources there's also this:
```json
    {
      "@id": "urn:solid-server:default:AuxiliaryStrategy",
      "@type": "RoutingAuxiliaryStrategy",
      "RoutingAuxiliaryStrategy:_sources": [
        {
          "@id": "urn:solid-server:default:AclStrategy"
        }
      ]
    },
```
So new strategies would them simply have to be added to this list to be supported by all relevant classes. Some classes only require the IdentifierStrategy parts, but the full strategy extends that interface so no need to have both.

RubenVerborgh commented on src/ldp/auxiliary/AuxiliaryStrategy.ts:
```diff
@@ -0,0 +1,42 @@
+   * Whether this auxiliary resource can be deleted when it's in a root storage container.
+   * @param identifier - Identifier of the auxiliary resource.
+   */
+  canDeleteRoot: (identifier: ResourceIdentifier) => boolean;
```
```suggestion
  requiresRootAuxiliary: (identifier: ResourceIdentifier) => boolean;
```

perhaps

RubenVerborgh commented on src/ldp/auxiliary/CompositeAuxiliaryStrategy.ts:
```diff
@@ -0,0 +1,58 @@
+ * An {@link AuxiliaryStrategy} that provides its functionality through the combination of
+ * an {@link AuxiliaryIdentifierStrategy}, {@link MetadataGenerator} and {@link Validator}.
+ */
+export class CompositeAuxiliaryStrategy implements AuxiliaryStrategy {
```
I just wonder if this is the right name, given that

> The composite pattern describes a group of objects that are treated the same way as a single instance of the same type of object.

Maybe

```suggestion
export class ComposedAuxiliaryStrategy implements AuxiliaryStrategy {
```

or so

RubenVerborgh commented on src/ldp/auxiliary/AuxiliaryStrategy.ts:
```diff
@@ -0,0 +1,42 @@
+   * Whether this auxiliary resource can be deleted when it's in a root storage container.
+   * @param identifier - Identifier of the auxiliary resource.
+   */
+  canDeleteRoot: (identifier: ResourceIdentifier) => boolean;
```
Also, can this be just a getter? Not a function?

joachimvh replied:
Then we can no longer pass the ResourceIdentifier parameter along, which is required for the composite strategy.

RubenVerborgh replied:
Woops, of course. Would still look into naming though, especially with the param.

RubenVerborgh commented on src/ldp/auxiliary/RdfValidator.ts:
```diff
@@ -0,0 +1,42 @@
+
+    const identifier = { path: representation.metadata.identifier.value };
+
+    // Read data in memory first so it does not get lost
+    const data = await arrayifyStream(representation.data);
+    const preferences = { type: { [INTERNAL_QUADS]: 1 }};
+
+    // Creating new representation since converter might edit metadata
+    const tempRepresentation = new BasicRepresentation(data, identifier, representation.metadata.contentType);
+    const result = await this.converter.handleSafe({ identifier, representation: tempRepresentation, preferences });
+    // Drain stream to make sure data was parsed correctly
+    await arrayifyStream(result.data);
```
In a later PR, let's create a utility for cloning a representation.
Perhaps mark this as TODO already.

But I'm missing a comment saying: we're just parsing the whole thing here.

I wonder whether we can double pipe (but I remember this causing a hanging issue at some point.

joachimvh replied:
Double piping is dangerous and can cause some weird errors if for example one of the pipes errors further down the line. Also backpressuring can get more complicated there. So I would prefer to avoid it as much as possible.

RubenVerborgh replied:
Agreed on double piping; but the risk is memory hogging. I would strongly suggest to make this a utility method, so we can at least keep track of where it is used.

RubenVerborgh commented on src/ldp/auxiliary/RoutingAuxiliaryIdentifierStrategy.ts:
```diff
@@ -0,0 +1,44 @@
+    this.sources = sources;
+  }
+
+  public getAuxiliaryIdentifier(): never {
```
Is the singular method ever used? Should we remove it?

joachimvh replied:
It is used by classes that need a specific auxiliary manager. There are some classes (acl authorizer, link metadata generator) that specifically need an acl based strategy, and those then use that function. They could also use the other one and then take the first element of the array of course, but that seems less nice.

RubenVerborgh commented on src/ldp/auxiliary/SuffixAuxiliaryIdentifierStrategy.ts:
```diff
@@ -0,0 +1,38 @@
+  protected readonly suffix: string;
+
+  public constructor(suffix: string) {
+    this.suffix = suffix;
```
Check if non-zero length?

RubenVerborgh reviewed (approved):
Thanks a lot. Just some random nits.

joachimvh commented on src/authorization/AuxiliaryAuthorizer.ts:
```diff
@@ -0,0 +1,50 @@
+  protected readonly logger = getLoggerFor(this);
+
+  private readonly resourceAuthorizer: Authorizer;
+  private readonly strategy: AuxiliaryIdentifierStrategy;
```
`strategy` is correct here since this authorizer handles all auxiliary resources. I do notice in the configs that currently it is linked to the acl strategy (which is sort of correct since it's the only one we have), but I should already link it to the composite one for when future auxiliary types get added.

RubenVerborgh replied:
I probably meant `auxStrategy`. Just found it weird to read `strategy` in the code below; like, what strategy?
